File: himura/src/main/java/be/vergauwen/simon/himura/MVPPresenter.java
package be.vergauwen.simon.himura;

import java.lang.ref.WeakReference;

public abstract class MVPPresenter<V extends MVPContract.View> implements MVPContract.Presenter<V> {
    private WeakReference<V> viewRef;

    @Override
    public V getView() {
        return viewRef == null ? null : viewRef.get();
    }

    @Override
    public void attachView(V view) {
        viewRef = new WeakReference<>(view);
    }

    @Override
    public void detachView() {
        if (viewRef != null) {
            viewRef.clear();
            viewRef = null;
        }
    }
}

File: app/src/main/kotlin/be/vergauwen/simon/mvp/MVPPresenter.kt
package be.vergauwen.simon.mvp

import java.lang.ref.WeakReference

abstract class MVPPresenter<V : MVPContract.View> : MVPContract.Presenter<V> {
  private var viewRef: WeakReference<MVPContract.View>? = null

  override fun getView(): V? = if (viewRef != null) viewRef!!.get() as V? else null

  override fun attachView(view: MVPContract.View) {
    viewRef = WeakReference(view)
  }

  override fun detachView() {
    viewRef?.clear()
    viewRef = null
  }
}
